File: clinic-referral-dashboard/node/src/webhookHandler.ts
/**
 * Framework-agnostic webhook orchestration: verify signature, persist
 * via EventStore, fan-out to the SSE broker, produce an outcome.  The
 * Express adapter translates the outcome into an HTTP response.
 */

import type { AppendArgs, EventRow } from './eventStore.js';
import { verifySignature } from './signatureVerifier.js';
import type { DashboardEvent, EventPublisher } from './sseBroker.js';

export type WebhookOutcome =
  | { readonly status: 'accepted'; readonly rowId: number }
  | { readonly status: 'bad_signature' }
  | { readonly status: 'bad_request'; readonly reason: string };

export interface AppendOnlyStore {
  append(args: AppendArgs): EventRow;
}

export interface Clock {
  nowIso(): string;
}

export interface WebhookHandlerOptions {
  readonly secret: string;
  readonly store: AppendOnlyStore;
  readonly publisher: EventPublisher;
  readonly clock: Clock;
}

export interface IncomingDelivery {
  readonly rawBody: Buffer;
  readonly signatureHeader: string | undefined;
  readonly eventTypeHeader: string | undefined;
}

export class WebhookHandler {
  constructor(private readonly options: WebhookHandlerOptions) {}

  handle(delivery: IncomingDelivery): WebhookOutcome {
    if (!verifySignature(delivery.rawBody, delivery.signatureHeader, this.options.secret)) {
      return { status: 'bad_signature' };
    }
    const bodyText = delivery.rawBody.toString('utf8');
    if (!isJsonObject(bodyText)) {
      return { status: 'bad_request', reason: 'Body is not a JSON object.' };
    }
    const eventType = (delivery.eventTypeHeader ?? '').trim();
    if (!eventType) {
      return { status: 'bad_request', reason: 'Missing X-Evagene-Event header.' };
    }
    const row = this.options.store.append({
      receivedAt: this.options.clock.nowIso(),
      eventType,
      body: bodyText,
    });
    this.options.publisher.publish(toDashboardEvent(row));
    return { status: 'accepted', rowId: row.id };
  }
}

function toDashboardEvent(row: EventRow): DashboardEvent {
  return {
    id: row.id,
    eventType: row.eventType,
    receivedAt: row.receivedAt,
    body: row.body,
  };
}

function isJsonObject(text: string): boolean {
  try {
    const parsed: unknown = JSON.parse(text);
    return typeof parsed === 'object' && parsed !== null && !Array.isArray(parsed);
  } catch {
    return false;
  }
}

File: webhook-audit-blotter/python/src/webhook_audit_blotter/webhook_handler.py
"""Framework-agnostic webhook orchestration."""

from __future__ import annotations

import json
from dataclasses import dataclass
from enum import Enum
from typing import Protocol

from .event_store import AppendArgs
from .signature_verifier import verify_signature


class OutcomeStatus(Enum):
    ACCEPTED = "accepted"
    BAD_SIGNATURE = "bad_signature"
    BAD_REQUEST = "bad_request"


@dataclass(frozen=True)
class WebhookOutcome:
    status: OutcomeStatus
    row_id: int | None = None
    reason: str | None = None


@dataclass(frozen=True)
class IncomingDelivery:
    raw_body: bytes
    signature_header: str | None
    event_type_header: str | None


class AppendOnlyStore(Protocol):
    def append(self, args: AppendArgs) -> int: ...


class Clock(Protocol):
    def now_iso(self) -> str: ...


class WebhookHandler:
    def __init__(self, *, secret: str, store: AppendOnlyStore, clock: Clock) -> None:
        self._secret = secret
        self._store = store
        self._clock = clock

    def handle(self, delivery: IncomingDelivery) -> WebhookOutcome:
        if not verify_signature(delivery.raw_body, delivery.signature_header, self._secret):
            return WebhookOutcome(status=OutcomeStatus.BAD_SIGNATURE)
        try:
            body_text = delivery.raw_body.decode("utf-8")
        except UnicodeDecodeError:
            return WebhookOutcome(status=OutcomeStatus.BAD_REQUEST, reason="Body is not UTF-8.")
        if not _is_json_object(body_text):
            return WebhookOutcome(
                status=OutcomeStatus.BAD_REQUEST,
                reason="Body is not a JSON object.",
            )
        event_type = (delivery.event_type_header or "").strip()
        if not event_type:
            return WebhookOutcome(
                status=OutcomeStatus.BAD_REQUEST,
                reason="Missing X-Evagene-Event header.",
            )
        row_id = self._store.append(
            AppendArgs(
                received_at=self._clock.now_iso(),
                event_type=event_type,
                body=body_text,
            ),
        )
        return WebhookOutcome(status=OutcomeStatus.ACCEPTED, row_id=row_id)


def _is_json_object(text: str) -> bool:
    try:
        parsed = json.loads(text)
    except ValueError:
        return False
    return isinstance(parsed, dict)
